### ISSUE TITLE
Collector causing w3wp crashes on Windows Server 2012 R2

### ISSUE THREAD

**Issue by georgiosd** [Microsoft/ApplicationInsights-dotnet-server#159]
Hello,

We're getting w3wp crashes on a Web API 2 project and using DebugDiag we found that they are coming from the App Insights modules. 

Any ideas what's causing this? We configured AppInsights from the wizard. 

Analysis from debug diag: https://www.dropbox.com/s/ey11kc8gvjkonci/MultipleDumps_CrashHangAnalysis.mht?dl=0

Thanks
Georgios

**Comment by SergeyKanzhelev** [comment]
From the first glance it looks like the host started unloading the application and sent `ThreadAboutException` to all threads. This exception crashed the Live Stream Metrics thread [here](https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/d69bd0e5f08766a9f045fbbe4bd42d984ed847d0/Src/PerformanceCollector/Shared/QuickPulseTelemetryModule.cs#L374). Since Sleep wasn't wrapped into `try{}catch` statement - it will cause the process to crash.

```
Type:   System.Threading.ThreadAbortException 
Message:   Thread was being aborted. 

[[HelperMethodFrame] (System.Threading.Thread.SleepInternal)] System.Threading.Thread.SleepInternal(Int32) 
mscorlib_ni!System.Threading.Thread.Sleep(Int32)+a 
mscorlib_ni!System.Threading.Thread.Sleep(System.TimeSpan)+36 
Microsoft.ApplicationInsights.Extensibility.PerfCounterCollector.QuickPulse.QuickPulseTelemetryModule.StateThreadWorker(System.Object)+28a 
mscorlib_ni!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)+160 
```

@tokaplan, can you please try to repro and fix it

**Comment by tokaplan** [comment]
@SergeyKanzhelev I'll look at the QuickPulse thread, you guys check InMemoryTelemetryChannel that seems to exhibit the same issue.

**Comment by tokaplan** [comment]
@georgiosd is this happening during a hosting environment shutdown? Do we expect to see those ThreadAbortException's in the first place?

**Comment by SergeyKanzhelev** [comment]
@tokaplan as far as I remember InMemoryChannel uses long running task. `ThreadAbortException` shouldn't crash the process even being unhandled.

**Comment by tokaplan** [comment]
@SergeyKanzhelev I'm seeing threads 13 and 17 in the crash report - they appear to be reported in the same way with the same stack, so if we consider one to be causing the crash, it appears that they should both be fixed. Am I reading the report wrong?

I'm still more interested in what caused the ThreadAbortException's in the first place.

**Comment by georgiosd** [comment]
It's actually very interesting. Once I disabled the app insights module I found that I shot myself in the foot.

Some time ago, I got into the habit of throwing `InvalidProgramException`s instead of the old-school `Trace.Assert` (which pops-up dialogs), not knowing that the exception is "special". It seems that when thrown, IIS will crash the app pool and try to restart the app. 

I haven't enabled back the module from this issue to check if that makes it work OK again.

Let me know if I can help any further.
